### PR TITLE
internal/config: refactor option parsing (#155)

### LINKED ISSUE #179: Allow proxy for https site running w/ self-signed certificate
**Is your feature request related to a problem?
It's not able to reverse proxy an internal https website use self-signed certificate.
The error is 6:36PM ERR http: proxy error: x509: certificate signed by unknown authority proxy=192.168.1.2:8006

**Describe the solution you'd like**
Add config parameter that ignores the certificate validation.

### LINKED ISSUE #204: internal/config: websockets feature switch should be route scoped
**Is your feature request related to a problem? Please describe.**

Currently, enabling web-socket support is all-or-nothing. 

**Describe the solution you'd like**

Websocket support be route based where other non-websocket based applications can be used together with routes with timeouts. Currently, if you enable websockets, other routes cannot use custom or default timeout handlers.

**Describe alternatives you've considered**

N/A.

**Explain any additional use-cases**

N/A 

**Additional context**

- #151 
- #77 
- #40

### PR DESCRIPTION
**Pull request by desimone**
- proxy: add self-signed certificate support. #179
- proxy: add skip tls certificate verification. #179
- proxy: Refactor websocket support to be route based. #204
- authorize: build whitelist from policy's URLs instead of strings.
- internal/httputil: merged httputil and https package.
- internal/config: merged config and policy packages.
- internal/metrics: removed unused measure struct.
- proxy/clients: refactor Addr fields to be urls.
- proxy: remove unused extend deadline function.
- proxy: use handler middleware for reverse proxy leg.
- proxy: change the way websocket requests are made (route based).

**General improvements**
- omitted value from range in several cases where for loop could be simplified.
- added error checking to many tests.
- standardize url parsing.
- remove unnecessary return statements. (see #201 )

**Related issues**
- Closes #204 
- Closes #179 
- Related to https://github.com/pomerium/pomerium/issues/155

This was a much more involved / comprehensive PR than I originally intended but many of the changes flowed naturally in trying to support #204 and #179 on a per route basis as well as to handle option immutability and scope. My fourth of july resolution is to have the rest of my patch sets be manageable in size and scope.   

**Config to test** 

```yaml
policy:
  - from: https://httpbin.corp.beyondperimeter.com
    to: http://localhost
    allowed_users:
      - bobbydesimone@gmail.com
      - bdd@pomerium.io

  - from: https://self-signed.corp.beyondperimeter.com
    to: https://self-signed.badssl.com
    allowed_users:
      - bobbydesimone@gmail.com
      - bdd@pomerium.io
    tls_custom_ca: "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURlVENDQW1HZ0F3SUJBZ0lKQUszMmhoR0JIcmFtTUEwR0NTcUdTSWIzRFFFQkN3VUFNR0l4Q3pBSkJnTlYKQkFZVEFsVlRNUk13RVFZRFZRUUlEQXBEWVd4cFptOXlibWxoTVJZd0ZBWURWUVFIREExVFlXNGdSbkpoYm1OcApjMk52TVE4d0RRWURWUVFLREFaQ1lXUlRVMHd4RlRBVEJnTlZCQU1NRENvdVltRmtjM05zTG1OdmJUQWVGdzB4Ck9UQTJNVEl4TlRNeE5UbGFGdzB5TVRBMk1URXhOVE14TlRsYU1HSXhDekFKQmdOVkJBWVRBbFZUTVJNd0VRWUQKVlFRSURBcERZV3hwWm05eWJtbGhNUll3RkFZRFZRUUhEQTFUWVc0Z1JuSmhibU5wYzJOdk1ROHdEUVlEVlFRSwpEQVpDWVdSVFUwd3hGVEFUQmdOVkJBTU1EQ291WW1Ga2MzTnNMbU52YlRDQ0FTSXdEUVlKS29aSWh2Y05BUUVCCkJRQURnZ0VQQURDQ0FRb0NnZ0VCQU1JRTdQaU03Z1RDczloUTFYQll6Sk1ZNjF5b2FFbXdJclg1bFo2eEt5eDIKUG16QVMyQk1UT3F5dE1BUGdMYXcrWExKaGdMNVhFRmRFeXQvY2NSTHZPbVVMbEEzcG1jY1lZejJRVUxGUnRNVwpoeWVmZE9zS25SRlNKaUZ6YklSTWVWWGswV3ZvQmoxSUZWS3RzeWpicXY5dS8yQ1ZTbmRyT2ZFazBURzIzVTNBCnhQeFR1VzFDcmJWOC9xNzFGZEl6U09jaWNjZkNGSHBzS09vM1N0L3FiTFZ5dEg1YW9oYmNhYkZYUk5zS0VxdmUKd3c5SGRGeEJJdUdhK1J1VDVxMGlCaWt1c2JwSkhBd25ucVA3aS9kQWNnQ3NrZ2paakZlRVU0RUZ5K2IrYTFTWQpRQ2VGeHhDN2MzRHZhUmhCQjBWVmZQbGtQejBzdzZsODY1TWFUSWJSeW9VQ0F3RUFBYU15TURBd0NRWURWUjBUCkJBSXdBREFqQmdOVkhSRUVIREFhZ2d3cUxtSmhaSE56YkM1amIyMkNDbUpoWkhOemJDNWpiMjB3RFFZSktvWkkKaHZjTkFRRUxCUUFEZ2dFQkFJaTV1OXc4bWdUNnBwQ2M3eHNHK0E5ZkkzVzR6K3FTS2FwaHI1bHM3MEdCS2JpWQpZTEpVWVpoUGZXcGgxcXRra1UwTEhGUG04M1ZhNTJlSUhyalhUMFZlNEt0TzFuMElBZkl0RmFXNjJDSmdoR1luCmp6dzByeXpnQzRQeUZwTk1uTnRCcm9QdS9iUGdXaU1nTE9OcEVaaGlneDRROHdmMVkvVTlzK3pDQ3hvSmxhS1IKTVhidVE4N1g3bS85VlJueHhvNk56NVpmN09USFRwTk9JNlZqYTBCeGJtSUFVNnlyaXc5VXJnaWJYZk9qM2o2bgpNVExCdWdVVklCMGJCYWFzSnNBTUsrdzRMQU52YXBlWjBET1NuT1I0S0syNEowT3lvRjVmSG1wNTllTTE3SW9GClFxQmh6cG1RVWd1bmVjRVc4QlRxck5wRzc5UjF1K1YrNHd3Y2tQYz0KLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo="

  # docker run -d -p 8080:8080 inanimate/echo-server
  - from: https://ws.corp.beyondperimeter.com
    to: http://localhost:8080
    allowed_users:
      - bobbydesimone@gmail.com
      - bdd@pomerium.io
    # tls_skip_verify
    allow_websockets: true

  - from: https://external-httpbin.corp.beyondperimeter.com
    to: https://httpbin.org
    allowed_users:
      - bobbydesimone@gmail.com
      - bdd@pomerium.io
```

**Checklist**:
- [x] updated docs
- [x] unit tests added
- [x] related issues referenced
- [x] updated CHANGELOG.md
- [x] ready for review

/cc @travisgroth @nareddyt

### CODE REVIEW

File: internal/config/options.go [hunk 1]
```diff
@@ -350,15 +297,15 @@ func (o *Options) parseHeaders() error {
...
 					// Something went wrong
-					return fmt.Errorf("Failed to decode headers environment variable from '%s'", o.HeadersEnv)
+					return fmt.Errorf("failed to decode headers from '%s'", o.HeadersEnv)
 				}
 			}
 
 		}
 		o.Headers = headers
-	} else if viper.IsSet("headers") {
+	} else if viper.IsSet("headers") && viper.Get("headers") != "" {
```
[the remark this answers is not in the document]
Comment (desimone) [reply internal/config/options.go]: Thanks -- I believe this was when I was trying to debug unit tests containing residual env vars which was ultimately solved by those `viper.Reset`s

File: internal/config/options.go [hunk 2]
```diff
@@ -128,219 +124,170 @@ type Options struct {
...
-	return o
+	return &o, nil
 }
 
 // OptionsFromViper builds the main binary's configuration
 // options by parsing environmental variables and config file
-func OptionsFromViper(configFile string) (Options, error) {
-	o := NewOptions()
-
+func OptionsFromViper(configFile string) (*Options, error) {
```
[the remark this answers is not in the document]
Comment (desimone) [reply internal/config/options.go]: This makes sense to me. Would it make sense to address in a follow up issue/PR as I believe (as you said) this chiefly effects the unit tests?

[the remark this answers is not in the document]
Comment (travisgroth) [reply internal/config/options.go]: Yep that's reasonable.

**Comment by travisgroth** [comment]
LGTM

File: internal/config/options.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (desimone) [reply internal/config/options.go]: See #213

**Pull request merged**